qiskit/pulse/ir/ir_instructions.py

```diff
@@ -0,0 +1,335 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2023.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+# pylint: disable=cyclic-import
+
+"""
+=========
+IR Instructions
+=========
+
+"""
+
+from typing import Optional
+from abc import ABC, abstractmethod
+
+import numpy as np
+
+from qiskit.pulse.exceptions import PulseError
+
+from qiskit.pulse import LogicalElement, Frame, SymbolicPulse, Waveform, Qubit, MemorySlot
+
+
+class BaseIRInstruction(ABC):
+    """Base class for PulseIR instruction
+
+    ``BaseIRInstruction`` serves as the base class for all instructions in PulseIR.
+    It sets common properties like ``duration``, ``initial_time``, and ``operand``.
+    """
+
+    @abstractmethod
+    def __init__(self, duration: int, operand, initial_time: Optional[int] = None):
```

Comment:
Because scheduling is such a major part of the compiler's job, I think it's not very productive to separate this issue. What is the benefit you see in separating this?